File: Data Structures & Algorithms/search-2d-matrix/submission-2.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows = len(matrix)
        cols = len(matrix[0])
        t,b = 0,rows-1
        r1=0
        while t<=b:
            mid=(t+b)//2
            if matrix[mid][0]==target:
                return True
            elif matrix[mid][0]<target:
                r1=mid
                t=mid+ 1
            else:
                b=mid-1
        l,r=0,cols-1
        while l<=r:
            mid = (l+r)//2
            if matrix[r1][mid]==target:
                return True
            elif matrix[r1][mid]<target:
                l=mid+ 1
            else:
                r=mid-1
        return False

File: Data Structures & Algorithms/unique-paths-ii/submission-2.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int rows = obstacleGrid.length;
        int cols = obstacleGrid[0].length;

        int[][] dp = new int[rows][cols];

        if (obstacleGrid[rows - 1][cols - 1] == 1 || obstacleGrid[0][0] == 1)
            return 0;
        dp[rows - 1][cols - 1] = 1;

        // last col
        for (int i = rows - 2; i > -1; i--) {
            if (obstacleGrid[i][cols - 1] == 1)
                dp[i][cols - 1] = 0;
            else
                dp[i][cols - 1] = dp[i + 1][cols - 1];
        }
        // last row
        for (int i = cols - 2; i > -1; i--) {
            if (obstacleGrid[rows - 1][i] == 1)
                dp[rows - 1][i] = 0;
            else
                dp[rows - 1][i] = dp[rows - 1][i + 1];
        }

        for (int row = rows - 2; row > -1; row--) {
            for (int col = cols - 2; col > -1; col--) {
                if (obstacleGrid[row][col] == 1)
                    dp[row][col] = 0;
                else
                    dp[row][col] = dp[row][col + 1] + dp[row + 1][col];
            }
        }

        return dp[0][0];
    }
}